src/join-8-auth.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+/// join-8.sol -- Non-standard token adapters
+
+// Copyright (C) 2018 Rain <rainbreak@riseup.net>
+// Copyright (C) 2018-2020 Maker Ecosystem Growth Holdings, INC.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.6.12;
+
+interface VatLike {
+    function slip(bytes32, address, int256) external;
+}
+
+interface GemLike {
+    function decimals() external view returns (uint8);
+    function transfer(address,uint256) external returns (bool);
+    function transferFrom(address,address,uint256) external returns (bool);
+    function erc20Impl() external view returns (address);
+}
+
+// AuthGemJoin8
+// For a token that has a lower precision than 18, has decimals and it is upgradable (like GUSD)
+
+contract AuthGemJoin8 {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike public immutable vat;
+    bytes32 public immutable ilk;
+    GemLike public immutable gem;
+    uint256 public immutable dec;
+    uint256 public live;  // Access Flag
+
+    // Events
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Join(address indexed usr, uint256 wad, address indexed msgSender);
+    event Exit(address indexed usr, uint256 wad);
+    event Cage();
+
+    mapping (address => uint256) public implementations;
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        gem = GemLike(gem_);
+        dec = GemLike(gem_).decimals();
+        require(GemLike(gem_).decimals() < 18, "AuthGemJoin8/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        setImplementation(GemLike(gem_).erc20Impl(), 1);
```

Comment:
I don't think you need the `GemLike` here:

```suggestion
        setImplementation(gem.erc20Impl(), 1);
```

Or is there preference for not using the stored `gem` in the constructor?

Reply:
This won't compile under `0.6.12`. Solidity doesn't let you read immutable variables in the constructor. This is fixed in later versions. It's a bit messy code-wise, but there is no gas overhead doing `address` casting like this.

Reply:
Not super-critical but I would consider giving the implementation as a ctr parameter, otherwise you can be "rugged" by its changing between what you last looked and the tx deployment. As similar code already exists with gem-join-8 I don't know if we should really change it but just raising it up for awareness.